### Overview

**SCRAM** is publicly available from https://github.com/akdess/scram

This vignette shows the basic steps for running scram.


### Installation

* Install latest version from GitHub (requires [devtools](https://github.com/hadley/devtools) package):

```r
if (!require("devtools")) {
  install.packages("devtools")
}
devtools::install_github("akdess/scram", dependencies = TRUE, build_vignettes = FALSE)
```

### Input data

The input to scram consists of raw expression matrix. Optionally (and preferably), the user can also provide Seurat R object with clusters and cluster markers. Cluster markers can be obtained by running Seurat R package function [FindAllMarkers](https://satijalab.org/seurat/reference/findallmarkers). 

Example dataset to run this this tutorial can be downloaded from [here].
example_seurat_object.rda : seurat object with clustering information
example_seurat_markers.rda : seurat cluster markers, if missing can be generated running the following code

```r
load("example_seurat_object.rda")
if(is.null(seuratObj$seurat_clusters)) meesage("seurat clusters are missing, please run FindClusters function in Seurat package")
combined.markers <- FindAllMarkers(seuratObj, only.pos = T, min.pct = 0.1, logfc.threshold = 0.5)
combined.markers <- combined.markers[combined.markers$p_val_adj<0.05,]
save("combined.markers", file="example_seurat_markers.rda")
```

Optionally, large scale CNV calls inferred from our R package [CaSpER] (https://github.com/akdess/CaSpER) can be provided as an input. SCRAM uses the matrix generated by [extractLargeScaleEvents](https://rpubs.com/akdes/673120) R function of CaSpER. SCRAM processes the large scale CNV call matrix using the following function: 

```r
library("scram")
library("CaSpER")
load("example_seurat_object.rda") 
load("example_finalChrMat.rda")
cnv <- readCASPER(finalChrMat=finalChrMat, seurat_obj=seuratObj, 
  amp_chr=c("chr7q","chr7p"), del_chr=c("chr10p","chr10q", "chr13q"), min_threshold=0, plot=T, project="example")
save("cnv", file="example_cnv.rda")
```
Glioma human and mouse cell type markers are manually curated for our scram package. 

Markers can be loaded through our SCRAM package
```r
library("scram")
data (markers_human)
data (markers_mouse)
```

### Running SCRAM

The main functions you will need to run SCRAM analysis are CreateSCRAMObject() and runSCRAM().
The SCRAM object is required for performing analysis on single-cell RNA-Seq. SCRAM object is created using following function

```r
load("example_seurat_object.rda") 
load("example_seurat_markers.rda")
load("example_cnv.rda")
data (markers_human)

scram_obj <- CreateSCRAMObject(seurat_obj=seuratObj, 
    expr_data = seuratObj@assays$RNA@data,
    cluster_markers=combined.markers, 
    required_markers=markers_human$required_markers, 
    supporting_markers=markers_human$supporting_markers, 
    rules=markers_human$rules,
    cell_type_markers=markers_human$cell_type_markers, 
    cnv=cnv, 
    organism="human") 

```

if you would like to run SCRAM without cnv calls, you can set cnv=NULL


```r
load("example_seurat_object.rda")
load("example_seurat_markers.rda")
data (markers_human)

scram_obj <- CreateSCRAMObject(seurat_obj=seuratObj, 
    expr_data = seuratObj@assays$RNA@data,
    cluster_markers=combined.markers, 
    required_markers=markers_human$required_markers, 
    supporting_markers=markers_human$supporting_markers, 
    rules=markers_human$rules,
    cell_type_markers=markers_human$cell_type_markers, 
    cnv=NULL, 
    organism="human") 

```

After creating the casper object we run frequent geneset analysis in different levels such as gene level, cell type level and etc. 

```r
scram_obj <- runSCRAM(object=scram_obj) 
```

SCRAM outputs the co-occuring genesets and celltypes, cell lineages etc as an excel file using the below function:

```r
writeSCRAMresults(object=scram_obj, project="example")
```

### Visualization

In order to comprehensively visualize the co-occuring genesets and celltypes, cell lineages etc, SCRAM generates plots using the below functions: 

```r
plotUMAPDirksCellClass(object=scram_obj, project="example")
plotUMAPSuvaCellClass(object=scram_obj, project="example")
plotUMAPCellClassSimple(object=scram_obj, project="example")
plotUMAPCellLineage(object=scram_obj, project="example")
plotUMAPCellTypePerCluster(object=scram_obj, project="example")
plotPercAnnotatedCells (object=scram_obj, project="example")
plotUMAPSuvaCellClass (object=scram_obj, project="example")
```


Tutorials
----------
1. scATACSeq human IDH-WT sample
[R code]()
2. scATACSeq humanIDH-Mutant sample
[R code]()
3. scRNASeq human IDH-WT sample
[R code]()
4. scRNASeq human IDH-Mutant sample
[R code]()
5. 10x human spatial transcriptomics dataset
[R code]()
6. scRNASeq mouse IDH-WT sample
[R code]()
7. scRNASeq mouse IDH-Mutant sample
[R code]()
