<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Identification of transdifferentiated immune and neuronal tumor cells in glioblastoma using frequent itemset approach">
<title> â€¢ scram</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="">
<meta property="og:description" content="Identification of transdifferentiated immune and neuronal tumor cells in glioblastoma using frequent itemset approach">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">scram</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level3">
<h3 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h3>
<p><strong>SCRAM</strong> is publicly available from <a href="https://github.com/akdess/scram" class="external-link uri">https://github.com/akdess/scram</a></p>
<p>This vignette shows the basic steps for running scram.</p>
</div>
<div class="section level3">
<h3 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h3>
<ul>
<li>Install latest version from GitHub (requires <a href="https://github.com/hadley/devtools" class="external-link">devtools</a> package):</li>
</ul>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://devtools.r-lib.org/" class="external-link">"devtools"</a></span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span>
<span class="op">}</span>
<span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org/reference/remote-reexports.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"akdess/scram"</span>, dependencies <span class="op">=</span> <span class="cn">TRUE</span>, build_vignettes <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="input-data">Input data<a class="anchor" aria-label="anchor" href="#input-data"></a>
</h3>
<p>The input to scram consists of raw expression matrix. Optionally (and preferably), the user can also provide Seurat R object with clusters and cluster markers. Cluster markers can be obtained by running Seurat R package function <a href="https://satijalab.org/seurat/reference/findallmarkers" class="external-link">FindAllMarkers</a>.</p>
<p>Example dataset to run this this tutorial can be downloaded from [here]. example_seurat_object.rda : seurat object with clustering information example_seurat_markers.rda : seurat cluster markers, if missing can be generated running the following code</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="st">"example_seurat_object.rda"</span><span class="op">)</span>
<span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NULL.html" class="external-link">is.null</a></span><span class="op">(</span><span class="va">seuratObj</span><span class="op">$</span><span class="va">seurat_clusters</span><span class="op">)</span><span class="op">)</span> <span class="fu">meesage</span><span class="op">(</span><span class="st">"seurat clusters are missing, please run FindClusters function in Seurat package"</span><span class="op">)</span>
<span class="va">combined.markers</span> <span class="op">&lt;-</span> <span class="fu">FindAllMarkers</span><span class="op">(</span><span class="va">seuratObj</span>, only.pos <span class="op">=</span> <span class="cn">T</span>, min.pct <span class="op">=</span> <span class="fl">0.1</span>, logfc.threshold <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span>
<span class="va">combined.markers</span> <span class="op">&lt;-</span> <span class="va">combined.markers</span><span class="op">[</span><span class="va">combined.markers</span><span class="op">$</span><span class="va">p_val_adj</span><span class="op">&lt;</span><span class="fl">0.05</span>,<span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/base/save.html" class="external-link">save</a></span><span class="op">(</span><span class="st">"combined.markers"</span>, file<span class="op">=</span><span class="st">"example_seurat_markers.rda"</span><span class="op">)</span></code></pre></div>
<p>Optionally, large scale CNV calls inferred from our R package [CaSpER] (<a href="https://github.com/akdess/CaSpER" class="external-link uri">https://github.com/akdess/CaSpER</a>) can be provided as an input. SCRAM uses the matrix generated by <a href="https://rpubs.com/akdes/673120" class="external-link">extractLargeScaleEvents</a> R function of CaSpER. SCRAM processes the large scale CNV call matrix using the following function:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"scram"</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"CaSpER"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="st">"example_seurat_object.rda"</span><span class="op">)</span> 
<span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="st">"example_finalChrMat.rda"</span><span class="op">)</span>
<span class="va">cnv</span> <span class="op">&lt;-</span> <span class="fu">readCASPER</span><span class="op">(</span>finalChrMat<span class="op">=</span><span class="va">finalChrMat</span>, seurat_obj<span class="op">=</span><span class="va">seuratObj</span>, 
  amp_chr<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/arules/man/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chr7q"</span>,<span class="st">"chr7p"</span><span class="op">)</span>, del_chr<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/arules/man/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chr10p"</span>,<span class="st">"chr10q"</span>, <span class="st">"chr13q"</span><span class="op">)</span>, min_threshold<span class="op">=</span><span class="fl">0</span>, plot<span class="op">=</span><span class="cn">T</span>, project<span class="op">=</span><span class="st">"example"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/save.html" class="external-link">save</a></span><span class="op">(</span><span class="st">"cnv"</span>, file<span class="op">=</span><span class="st">"example_cnv.rda"</span><span class="op">)</span></code></pre></div>
<p>Glioma human and mouse cell type markers are manually curated for our scram package.</p>
<p>Markers can be loaded through our SCRAM package</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"scram"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span> <span class="op">(</span><span class="va">markers_human</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span> <span class="op">(</span><span class="va">markers_mouse</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="running-scram">Running SCRAM<a class="anchor" aria-label="anchor" href="#running-scram"></a>
</h3>
<p>The main functions you will need to run SCRAM analysis are CreateSCRAMObject() and runSCRAM(). The SCRAM object is required for performing analysis on single-cell RNA-Seq. SCRAM object is created using following function</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="st">"example_seurat_object.rda"</span><span class="op">)</span> 
<span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="st">"example_seurat_markers.rda"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="st">"example_cnv.rda"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span> <span class="op">(</span><span class="va">markers_human</span><span class="op">)</span>

<span class="va">scram_obj</span> <span class="op">&lt;-</span> <span class="fu">CreateSCRAMObject</span><span class="op">(</span>seurat_obj<span class="op">=</span><span class="va">seuratObj</span>, 
    expr_data <span class="op">=</span> <span class="va">seuratObj</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">@</span><span class="va">data</span>,
    cluster_markers<span class="op">=</span><span class="va">combined.markers</span>, 
    required_markers<span class="op">=</span><span class="va">markers_human</span><span class="op">$</span><span class="va">required_markers</span>, 
    supporting_markers<span class="op">=</span><span class="va">markers_human</span><span class="op">$</span><span class="va">supporting_markers</span>, 
    rules<span class="op">=</span><span class="va">markers_human</span><span class="op">$</span><span class="va">rules</span>,
    cell_type_markers<span class="op">=</span><span class="va">markers_human</span><span class="op">$</span><span class="va">cell_type_markers</span>, 
    cnv<span class="op">=</span><span class="va">cnv</span>, 
    organism<span class="op">=</span><span class="st">"human"</span><span class="op">)</span> </code></pre></div>
<p>if you would like to run SCRAM without cnv calls, you can set cnv=NULL</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="st">"example_seurat_object.rda"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="st">"example_seurat_markers.rda"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span> <span class="op">(</span><span class="va">markers_human</span><span class="op">)</span>

<span class="va">scram_obj</span> <span class="op">&lt;-</span> <span class="fu">CreateSCRAMObject</span><span class="op">(</span>seurat_obj<span class="op">=</span><span class="va">seuratObj</span>, 
    expr_data <span class="op">=</span> <span class="va">seuratObj</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">@</span><span class="va">data</span>,
    cluster_markers<span class="op">=</span><span class="va">combined.markers</span>, 
    required_markers<span class="op">=</span><span class="va">markers_human</span><span class="op">$</span><span class="va">required_markers</span>, 
    supporting_markers<span class="op">=</span><span class="va">markers_human</span><span class="op">$</span><span class="va">supporting_markers</span>, 
    rules<span class="op">=</span><span class="va">markers_human</span><span class="op">$</span><span class="va">rules</span>,
    cell_type_markers<span class="op">=</span><span class="va">markers_human</span><span class="op">$</span><span class="va">cell_type_markers</span>, 
    cnv<span class="op">=</span><span class="cn">NULL</span>, 
    organism<span class="op">=</span><span class="st">"human"</span><span class="op">)</span> </code></pre></div>
<p>After creating the casper object we run frequent geneset analysis in different levels such as gene level, cell type level and etc.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scram_obj</span> <span class="op">&lt;-</span> <span class="fu">runSCRAM</span><span class="op">(</span>object<span class="op">=</span><span class="va">scram_obj</span><span class="op">)</span> </code></pre></div>
<p>SCRAM outputs the co-occuring genesets and celltypes, cell lineages etc as an excel file using the below function:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">writeSCRAMresults</span><span class="op">(</span>object<span class="op">=</span><span class="va">scram_obj</span>, project<span class="op">=</span><span class="st">"example"</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="visualization">Visualization<a class="anchor" aria-label="anchor" href="#visualization"></a>
</h3>
<p>In order to comprehensively visualize the co-occuring genesets and celltypes, cell lineages etc, SCRAM generates plots using the below functions:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">plotUMAPDirksCellClass</span><span class="op">(</span>object<span class="op">=</span><span class="va">scram_obj</span>, project<span class="op">=</span><span class="st">"example"</span><span class="op">)</span>
<span class="fu">plotUMAPSuvaCellClass</span><span class="op">(</span>object<span class="op">=</span><span class="va">scram_obj</span>, project<span class="op">=</span><span class="st">"example"</span><span class="op">)</span>
<span class="fu">plotUMAPCellClassSimple</span><span class="op">(</span>object<span class="op">=</span><span class="va">scram_obj</span>, project<span class="op">=</span><span class="st">"example"</span><span class="op">)</span>
<span class="fu">plotUMAPCellLineage</span><span class="op">(</span>object<span class="op">=</span><span class="va">scram_obj</span>, project<span class="op">=</span><span class="st">"example"</span><span class="op">)</span>
<span class="fu">plotUMAPCellTypePerCluster</span><span class="op">(</span>object<span class="op">=</span><span class="va">scram_obj</span>, project<span class="op">=</span><span class="st">"example"</span><span class="op">)</span>
<span class="fu">plotPercAnnotatedCells</span> <span class="op">(</span>object<span class="op">=</span><span class="va">scram_obj</span>, project<span class="op">=</span><span class="st">"example"</span><span class="op">)</span>
<span class="fu">plotUMAPSuvaCellClass</span> <span class="op">(</span>object<span class="op">=</span><span class="va">scram_obj</span>, project<span class="op">=</span><span class="st">"example"</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="tutorials">Tutorials<a class="anchor" aria-label="anchor" href="#tutorials"></a>
</h2>
<ol style="list-style-type: decimal">
<li>scATACSeq human IDH-WT sample <a href="">R code</a>
</li>
<li>scATACSeq humanIDH-Mutant sample <a href="">R code</a>
</li>
<li>scRNASeq human IDH-WT sample <a href="">R code</a>
</li>
<li>scRNASeq human IDH-Mutant sample <a href="">R code</a>
</li>
<li>10x human spatial transcriptomics dataset <a href="">R code</a>
</li>
<li>scRNASeq mouse IDH-WT sample <a href="">R code</a>
</li>
<li>scRNASeq mouse IDH-Mutant sample <a href="">R code</a>
</li>
</ol>
</div>

  </main><aside class="col-md-3"><div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li>NA</li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing scram</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li><br><small class="roles"> Maintainer </small>  </li>
<li><a href="authors.html">More about authors...</a></li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Akdes Serin Harmanci.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
